File: web/options/options.dart
/*!
 * Copyright (C) 2012-2013, The SwitchyOmega Authors. Please see the AUTHORS file
 * for details.
 *
 * This file is part of SwitchyOmega.
 *
 * SwitchyOmega is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SwitchyOmega is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SwitchyOmega.  If not, see <http://www.gnu.org/licenses/>.
 */

library switchy_options;

import 'dart:html';
import 'dart:json' as JSON;
import 'dart:async';
import 'package:web_ui/web_ui.dart';
import 'package:crypto/crypto.dart';
import 'package:switchyomega/switchyomega.dart';
import 'package:switchyomega/html/converters.dart' as convert;
import 'package:switchyomega/browser/lib.dart';
import 'package:switchyomega/browser/message/lib.dart';
import 'package:switchyomega/communicator.dart';
import 'editors.dart';

@observable
List<Profile> modalCannotDeleteProfile_referring = toObservable([]);
@observable
Profile modalDeleteProfile_profile = null;

@observable
String modalRenameProfile_newName = null;
@observable
String modalRenameProfile_oldName = null;

@observable
Profile modalDeleteRule_profile = null;
@observable
Rule modalDeleteRule_rule = null;

@observable
Profile modalResetRules_profile = null;
@observable
Profile modalResetRules_resultProfile = null;

@observable
String modalNewProfile_name = '';

@observable
String currentProfileName = '';

void deleteProfile(Profile profile) {
  options.profiles.remove(profile);
  if (currentProfileName == profile.name) {
    currentProfileName = new DirectProfile().name;
  }
  if (options.startupProfileName == profile.name) {
    options.startupProfileName = null;
  }
  options.quickSwitchProfiles.remove(profile.name);

  if (profile is FixedProfile) {
    fixedProfileEditors.remove(profile);
  } else if (profile is SwitchProfile) {
    (profile as SwitchProfile).forEach((rule) => ruleEditors.remove(rule));
  }
}

void requestProfileDelete(Profile profile) {
  deliverChangesSync();
  var ref = profile is IncludableProfile ?
      options.profiles.referredBy(profile) : [];
  if (!ref.isEmpty) {
    // Cannot delete profile.
    modalCannotDeleteProfile_referring = ref.toList();
    js.send('modal.profile.cannotDelete', null, (_, [reply]) {
      // Modal closed.
      modalCannotDeleteProfile_referring = [];
    });
  } else {
    modalDeleteProfile_profile = profile;
    js.send('modal.profile.delete', null, (action, [reply]) {
      if (action == 'delete') {
        deleteProfile(profile);
        js.send('tab.set');
      }
      modalDeleteProfile_profile = null;
    });
  }
}

void requestProfileRename(Profile profile) {
  modalRenameProfile_newName = modalRenameProfile_oldName = profile.name;
  js.send('modal.profile.rename', null, (action, [reply]) {
    if (action == 'rename') {
      if (modalRenameProfile_newName != modalRenameProfile_oldName) {
        options.profiles.renameProfile(modalRenameProfile_oldName,
            modalRenameProfile_newName);
        js.send('tab.set', '#profile-$modalRenameProfile_newName');
      }
    }
    modalRenameProfile_newName = modalRenameProfile_oldName = null;
  });
}

bool modalRenameProfile_isValid() {
  if (modalRenameProfile_newName == modalRenameProfile_oldName) return true;
  if (modalRenameProfile_newName.isEmpty) return false;
  if (options.profiles[modalRenameProfile_newName] != null) return false;
  return true;
}

bool modalNewProfile_isValid() {
  if (modalNewProfile_name.isEmpty) return false;
  if (options.profiles[modalNewProfile_name] != null) return false;
  return true;
}

String bypassListToText(List<BypassCondition> list) =>
    list.map((b) => b.pattern).join('\n');

void handleFixedServerUI() {
  dynamicEvent('change', '.bypass-list',  (e, TextAreaElement bypassList) {
    var profile_name = closestElement(bypassList, '.tab-pane')
        .attributes['data-profile'];
    var profile = options.profiles[profile_name] as FixedProfile;
    profile.bypassList.clear();
    profile.bypassList.addAll(bypassList.value.split('\n')
        .map((l) => l.trim()).where((l) => !l.isEmpty)
        .map((l) => new BypassCondition(l)));
    // Update the text in the textarea.
    bypassList.value = bypassListToText(profile.bypassList);
  });
}

void handleSwitchProfileUI() {
  new EventStreamProvider('x-sort').forTarget(document).listen((Event e) {
    var target = e.target as Element;
    var profile_name = closestElement(target, '.tab-pane')
        .attributes['data-profile'];
    var profile = options.profiles[profile_name] as SwitchProfile;
    var index_old = int.parse(target.attributes['data-index-old']);
    var index_new = int.parse(target.attributes['data-index-new']);
    var rule = profile.removeAt(index_old);
    profile.insert(index_new, rule);
  });
}

void moveRuleUp(SwitchProfile profile, Rule rule) {
  var index_old = profile.indexOf(rule);
  if (index_old > 0) {
    profile.removeAt(index_old);
    profile.insert(index_old - 1, rule);
  }
}

void moveRuleDown(SwitchProfile profile, Rule rule) {
  var index_old = profile.indexOf(rule);
  if (index_old < profile.length - 1) {
    profile.removeAt(index_old);
    profile.insert(index_old + 1, rule);
  }
}

void requestRemoveRule(SwitchProfile profile, Rule rule) {
  if (options.confirmDeletion) {
    modalDeleteRule_profile = profile;
    modalDeleteRule_rule = rule;
    js.send('modal.rule.delete', null, (action, [reply]) {
      if (action == 'delete') {
        removeRule(profile, rule);
      }
      modalDeleteRule_profile = null;
      modalDeleteRule_rule = null;
    });
  } else {
    removeRule(profile, rule);
  }
}

void removeRule(SwitchProfile profile, Rule rule) {
  profile.remove(rule);
  ruleEditors.remove(rule);
}

List<Profile> validResultProfilesFor(InclusiveProfile profile) {
  return options.profiles.validResultProfilesFor(profile).toList();
}

void addRule(SwitchProfile profile) {
  var condition = new HostWildcardCondition('*.example.com');
  var profileName = profile.length > 0 ?
      profile.last.profileName : profile.defaultProfileName;

  profile.add(new Rule(condition, profileName));
}

void resetRules(SwitchProfile profile) {
  profile.forEach((rule) {
    rule.profileName = profile.defaultProfileName;
  });
}

void requestResetRules(SwitchProfile profile) {
  modalResetRules_profile = profile;
  modalResetRules_resultProfile = options.profiles[profile.defaultProfileName];
  js.send('modal.rule.reset', null, (action, [reply]) {
    if (action == 'reset') {
      resetRules(profile);
    }
    modalResetRules_profile = null;
    modalResetRules_resultProfile = null;
  });
}

Communicator safe = new Communicator(window.top);
Communicator js = new Communicator(window);
Browser browser = new MessageBrowser(safe);

@observable
StoredSwitchyOptions options = null;

List<String> profileTypes = ['FixedProfile',
                             'SwitchProfile',
                             'PacProfile',
                             'SwitchyRuleListProfile',
                             'AutoProxyRuleListProfile'];

Map<String, String> profileIcons = {
  'FixedProfile': 'icon-globe',
  'PacProfile': 'icon-tasks',
  'RulelistProfile': 'icon-list',
  'SwitchProfile': 'icon-retweet',
  'SwitchyRuleListProfile': 'icon-list',
  'AutoProxyRuleListProfile': 'icon-list'
};

Map<FixedProfile, FixedProfileEditor> fixedProfileEditors =
    new Map<FixedProfile, FixedProfileEditor>();

Map<Rule, RuleEditor> ruleEditors = new Map<Rule, RuleEditor>();

void exportPac() {
  var current = options.profiles[currentProfileName];
  if (current is InclusiveProfile) {
    js.send('file.saveAs', {
      'name': 'SwitchyOmega.pac',
      'content': current.toScript()
    });
  }
}

void exportOptions() {
  js.send('file.saveAs', {
    'name': 'SwitchyOmegaOptions.json',
    'content': JSON.stringify(options)
  });
}

void restoreOptions(String data) {
  StoredSwitchyOptions newOptions = null;

  data = data.trim();
  if (data[0] != '{') {
    Map<String, Object> json = null;
    try {
      data = new String.fromCharCodes(CryptoUtils.base64StringToBytes(data));
      json = JSON.parse(data);
    } catch (e) {

      query('#options-import-format-error').style.top = "0";
      return;
    }
    newOptions = upgradeOptions(json, browser.storage);
  } else {
    try {
      newOptions = new StoredSwitchyOptions.fromPlain(JSON.parse(data),
          browser.storage);
    } catch (e) {
      query('#options-import-format-error').style.top = "0";
      return;
    }
  }
  ChangeUnobserver unobserve;
  unobserve = observe(() => options, (_) {
    query('#options-import-success').style.top = "0";
    unobserve();
  });
  options = newOptions;
  options.storeAll();
}

void restoreLocal(FileUploadInputElement file) {
  if (file.files.length > 0 && file.files[0].name.length > 0) {
    var r = new FileReader();
    r.onLoad.listen((_) {
      restoreOptions(r.result as String);
    });
    r.readAsText(file.files[0]);
    file.value = "";
  }
}

void restoreOnline() {
  var button = query('#restore-online') as ButtonElement;
  button.disabled = true;
  var url = (query('#restore-url') as InputElement).value;
  browser.download(url).then((data) {
    restoreOptions(data);
  }).catchError((e) {
    query('#options-import-download-error').style.top = "0";
  }).whenComplete(() {
    button.disabled = false;
  });
}

void saveOptions(Event e) {
  e.preventDefault();
  safe.send('options.set', JSON.stringify(options), (_, [__]) {
    query('#options-save-success').style.top = "0";
  });
}

void handleNewProfileUI() {
  js.on('profile.create', (String type, [Function respond]) {
    Profile p = null;
    switch (type) {
      case 'FixedProfile':
        p = new FixedProfile(modalNewProfile_name);
        break;
      case 'PacProfile':
        p = new PacProfile(modalNewProfile_name);
        break;
      case 'SwitchProfile':
        p = new SwitchProfile(modalNewProfile_name, new DirectProfile().name);
        break;
      case 'SwitchyRuleListProfile':
        p = new SwitchyRuleListProfile(modalNewProfile_name,
            new DirectProfile().name, new DirectProfile().name);
        break;
      case 'AutoProxyRuleListProfile':
        p = new AutoProxyRuleListProfile(modalNewProfile_name,
            new DirectProfile().name, new DirectProfile().name);
        break;
      default:
        throw new UnimplementedError();
    }
    if (p != null) {
      options.profiles.add(p);

      js.send('tab.set', '#profile-$modalNewProfile_name');
    }
    modalNewProfile_name = '';
  });
}

Map<String, Object> optionsBackup = null;

void handleActionsUI() {
  js.on('options.undo', (String type, [Function respond]) {
    var o = optionsBackup;
    optionsBackup = {
                     'options': options.toJson(),
                     'currentProfileName': currentProfileName,
                     'tab': null
    };
    ChangeUnobserver unobserve;
    unobserve = observe(() => options, (_) {
      js.send('tab.set', o['tab']);
      query('#options-undo-success').style.top = "0";
      unobserve();
    });
    options = new StoredSwitchyOptions.fromPlain(o['options'],
        browser.storage);
    options.storeAll();

    currentProfileName = o['currentProfileName'];
  });

  js.on('options.reset', (String type, [Function respond]) {
    safe.send('options.reset', null,
        (Map<String, Object> o, [Function respond]) {
      ChangeUnobserver unobserve;
      unobserve = observe(() => options, (_) {
        js.send('tab.set', o['tab']);
        query('#options-reset-success').style.top = "0";
        unobserve();
      });
      options = new StoredSwitchyOptions.fromPlain(o['options'],
          browser.storage);
      currentProfileName = o['currentProfileName'];
    });
  });
}

@observable List<Profile> quickSwitchDisabledProfiles = [];

void handleQuickSwitchUI() {
  js.on('quickswitch.update', (String type, [Function respond]) {
    options.quickSwitchProfiles.clear();
    options.quickSwitchProfiles.addAll(
        queryAll('#cycle-enabled li .profile-name').map((e) => e.text));
  });
  var refresh = (_) {
      var set = new Set<String>.from(options.profiles.map((p) => p.name));
      set.removeAll(options.quickSwitchProfiles);
      quickSwitchDisabledProfiles = set.map(options.getProfileByName)
          .toList(growable: false);
      js.send('quickswitch.refresh');
  };
  observe(() => options.quickSwitchProfiles, refresh);
  observe(() => options.profiles, refresh);
  refresh(null);
}

void downloadProfileNow(UpdatingProfile p, Event e) {
  var button = e.target as ButtonElement;
  button.disabled = true;
  browser.download(p.updateUrl).then((data) {
    p.applyUpdate(data);
    query('#options-profile-download-success').style.top = "0";
  }).catchError((e) {
    query('#options-profile-download-error').style.top = "0";
  }).whenComplete(() {
    button.disabled = false;
    safe.send('options.set', JSON.stringify(options));
  });
}

void main() {
  safe.send('options.get', null, (Map<String, Object> o, [Function respond]) {
    optionsBackup = o;
    ChangeUnobserver unobserve;
    unobserve = observe(() => options, (_) {
      js.send('tab.set', o['tab']);
      handleQuickSwitchUI();
      unobserve();
    });

    observe(() => options, (_) {
      if (options != null) {
        js.send('options.init');
        js.send('tab.set');
      }
    });

    options = new StoredSwitchyOptions.fromPlain(o['options'],
        browser.storage);
    currentProfileName = o['currentProfileName'];
  });

  handleActionsUI();
  handleNewProfileUI();
  handleFixedServerUI();
  handleSwitchProfileUI();
  autoBindToDataList(document.documentElement);
}

File: web/options/options.js
/*!
 * Copyright (C) 2012-2013, The SwitchyOmega Authors. Please see the AUTHORS
 * file for details.
 *
 * This file is part of SwitchyOmega.
 *
 * SwitchyOmega is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SwitchyOmega is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SwitchyOmega.  If not, see <http://www.gnu.org/licenses/>.
 */

(function () {
  'use strict';
  /* jshint jquery: true */
  /* global chrome: false, Communicator: false, strings: false */
  /* global i18nDict: false, i18nTemplate: false */

  // Dart uses symbol $ in generated scripts.
  jQuery.noConflict();
  var $ = jQuery;

  // document.register is renamed to document.registerElement recently.
  document.register = document.register || document.registerElement;

  var storage = chrome.storage.local;

  var dart = new Communicator(window);

  var i18nCache = {};
  strings.forEach(function (name) {
    i18nCache[name] = chrome.i18n.getMessage(name);
  });
  var i18n = new i18nDict(i18nCache); // jshint ignore:line
  window.i18n = new i18nDict(i18nCache); // jshint ignore:line

  var MutationObserver = window.MutationObserver ||
    window.WebKitMutationObserver;
  var ob = new MutationObserver(function (mutations) {
    mutations.forEach(function (record) {
      switch (record.type) {
        case 'attributes':
          i18nTemplate.process(record.target, i18n);
          break;
        case 'childList':
          for (var i = 0; i < record.addedNodes.length; i++) {
            var ele = record.addedNodes[i];
            if (ele instanceof Element) {
              i18nTemplate.process(ele, i18n);
            }
          }
          break;
      }
    });
  });
  ob.observe(document, {
    childList: true,
    attributes: true,
    subtree: true
  });
  window.onShadowHostReady = function (shadowHost) {
    var shadowRoot = shadowHost.shadowRoot || shadowHost.webkitShadowRoot ||
        shadowHost.mozShadowRoot;
    // Use shadowHost in case that Shadow DOM is being polyfilled.
    shadowRoot = shadowRoot || shadowHost;
    i18nTemplate.process(shadowRoot, i18n);
    ob.observe(shadowRoot, {
      childList: true,
      attributes: true,
      subtree: true
    });
  };
  i18nTemplate.process(document, i18n);

  window.bsShowModal = function (modal) {
    modal = $(modal);
    if (!modal.data('bs-modal-handler')) {
      modal.on('hide', function () {
        modal[0].dispatchEvent(new CustomEvent('modalhide'));
      });
      modal.data('bs-modal-handler', 'installed');
    }
    modal.modal('show');
  };
  window.bsHideModal = function (modal) {
    $(modal).modal('hide');
  };

  var lastTab = location.hash || localStorage['options_last_tab'];
  var loadOptions = function (respond) {
    storage.get(null, function (items) {
      respond({
        'options': items,
        'tab': localStorage['options_last_tab'],
        'currentProfileName': localStorage['currentProfileName'] || 'direct'
      });
    });
  };

  dart.on({
    'options.init': function () {
      // Sortable
      var containers = $('.cycle-profile-container');
      containers.sortable({
        connectWith: '.cycle-profile-container',
        tolerance: 'pointer',
        axis: 'y',
        forceHelperSize: true,
        forcePlaceholderSize: true
      }).disableSelection();
      $('#cycle-enabled').sortable({
        update: function () {
          dart.send('quickswitch.update');
        }
      });
      containers.data('cycleSortable', 'true');

      // Memorize Tab
      $('#options-nav').on('shown', 'a[data-toggle="tab"]', function (e) {
        lastTab = e.target.getAttribute('href');
        localStorage['options_last_tab'] = lastTab;
      });

      // Restore options from local files.
      $('#restore-local').on('click', function () {
        $('#restore-local-file').click();
      });
    },
    'quickswitch.refresh': function () {
      var containers = $('.cycle-profile-container');
      if (containers.data('cycleSortable')) {
        containers.sortable('refresh');
      }
    },
    'tab.set': function (tabhref) {
      tabhref = tabhref || lastTab;
      var tabs = $('#options-nav a[data-toggle="tab"]');
      var tab = null;
      if (tabhref) {
        tabs.each(function (_, e) {
          if (e.getAttribute('href') === tabhref) {
            tab = $(e);
            lastTab = tabhref;
          }
        });
      }
      if (tab === null) {
        tab = tabs.first();
      }
      tab.tab('show');
    },
    'file.saveAs': function (data, reply) {
      /* global saveAs: false */
      saveAs(new Blob([data.content]), data.name);
      reply();
    },
    'ajax.get': function (url, respond) {
      jQuery.ajax({
        url: url,
        cache: false,
        dataType: 'text',
        success: function (data) {
          respond({'data': data});
        },
        error: function (_, status, error) {
          respond({'status': status, 'error': error});
        }
      });
    },
    'error.log': function (data) {
      window.onerror(data.message, data.url, data.line);
    },
    'options.get': function (data, respond) {
      loadOptions(respond);
    },
    'options.reset': function (data, respond) {
      chrome.runtime.sendMessage({
        action: 'options.reset'
      }, function () {
        loadOptions(respond);
      });
    },
    'storage.get': function (keys, respond) {
      storage.get(keys, respond);
    },
    'storage.set': function (items, respond) {
      storage.set(items, respond);
    },
    'storage.remove': function (keys, respond) {
      if (keys == null) {
        storage.clear(respond);
      } else {
        storage.remove(keys, respond);
      }
    }
  });

  $(document).ready(function () {
    // Conditions Sort
    var MutationObserver = window.MutationObserver ||
                           window.WebKitMutationObserver;
    new MutationObserver(function (mutations) {
      var handleConditions = function (_, e) {
        var tb = $(e);
        tb.disableSelection().sortable({
          handle: '.sort-bar',
          tolerance: 'pointer',
          axis: 'y',
          forceHelperSize: true,
          forcePlaceholderSize: true,
          containment: 'parent',
          start: function (e, ui) {
            ui.item.data('index-old', ui.item.index());
          },
          update: function (e, ui) {
            ui.item.attr('data-index-old', ui.item.data('index-old'));
            ui.item.attr('data-index-new', ui.item.index());
            var evt = document.createEvent('CustomEvent');
            evt.initCustomEvent('x-sort', true, false, null);
            ui.item[0].dispatchEvent(evt);

            ui.item.removeAttr('data-index-old');
            ui.item.removeAttr('data-index-new');
          }
        });
      };
      mutations.forEach(function (record) {
        switch (record.type) {
          case 'childList':
            for (var i = 0; i < record.addedNodes.length; i++) {
              if (record.addedNodes[i] instanceof Element) {
                $('.conditions', record.addedNodes[i]).each(handleConditions);
              }
            }
            break;
        }
      });
    }).observe(document, {
      childList: true,
      subtree: true
    });

    var fireInputAndChangeEvent = function (target) {
      var evt = document.createEvent('HTMLEvents');
      evt.initEvent('input', true, false);
      target.dispatchEvent(evt);

      evt = document.createEvent('HTMLEvents');
      evt.initEvent('change', true, true);
      target.dispatchEvent(evt);
    };

    // Click anywhere to close alert.
    $(document).on('click', function () {
      $('#alert')[0].dispatchEvent(new CustomEvent('alerthide'));
    });

    // Clear input
    $('body').on('click', '.clear-input', function (e) {
      var button = $(e.target).closest('.clear-input');
      var input = button.prev('input');
      if (button.hasClass('revert')) {
        input.val(input.attr('data-restore'));
        fireInputAndChangeEvent(input[0]);
        button.removeClass('revert');
        button.find('i').removeClass('icon-repeat').addClass('icon-remove');
      } else {
        var v = input.val();
        if (v) {
          input.attr('data-restore', v);
          input.val('');
          fireInputAndChangeEvent(input[0]);
          button.addClass('revert');
          button.find('i').removeClass('icon-remove').addClass('icon-repeat');
          var handler = function () {
            if (input.val()) {
              button.removeClass('revert');
              $('i', button).removeClass('icon-repeat').addClass('icon-remove');
              input.off(eventMap);
            }
          };
          var eventMap = {'change': handler, 'keyup': handler,
            'keydown': handler};
          input.on(eventMap);
        }
      }
    });
  });

})();
